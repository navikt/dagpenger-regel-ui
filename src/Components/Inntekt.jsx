import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Element, EtikettLiten } from 'nav-frontend-typografi';
import Ekspanderbartpanel from 'nav-frontend-ekspanderbartpanel';
import { Input } from 'nav-frontend-skjema';
import { Knapp } from 'nav-frontend-knapper';
import { Field } from 'formik';

const sumInntekter = inntekter => inntekter.reduce((acc, val) => acc + val.beloep, 0);

const formatertPengesum = tall => new Intl.NumberFormat('nb-NO', { style: 'currency', currency: 'NOK' }).format(tall);

const InputField = ({
  beskrivelse, field, form, ...props
}) => (<Input label={beskrivelse} type="number" {...field} value={field.value} />);

const ReadOnlyField = ({
  beskrivelse, field, form, ...props
}) => (
  <div>
    <EtikettLiten>{beskrivelse}</EtikettLiten>
    <Element className="inntektbeloep">{formatertPengesum(field.value)}</Element>
  </div>
);

const Inntekt = ({
  readOnly, inntekter, rowId, columnId, monthIndex
}) => {
  const [editMode, setEditMode] = useState(true);

  return (
    <div className={`item inntekter inntekter--${rowId}--${columnId}`}>
      <Ekspanderbartpanel
        tittel={formatertPengesum(sumInntekter(inntekter.filter(inntekt => inntekt.virksomhet.identifikator === rowId)))}
        tittelProps="element"
      >
        {!readOnly && <Knapp htmlType="button" onClick={() => setEditMode(!editMode)}>Rediger</Knapp>}
        {inntekter.map((inntekt, index) => (
          inntekt.virksomhet.identifikator === rowId &&
          <div key={inntekt.beskrivelse} className="inntekt">
            <Field beskrivelse={inntekt.beskrivelse} name={`inntekt.arbeidsInntektMaaned[${monthIndex}].arbeidsInntektInformasjon.inntektListe[${index}].beloep`} component={editMode ? ReadOnlyField : InputField} />
          </div>
        ))}
      </Ekspanderbartpanel>
    </div>
  );
};

Inntekt.propTypes = {
  inntekter: PropTypes.arrayOf(PropTypes.shape({
    header: PropTypes.string,
    beskrivelse: PropTypes.string,
    beloep: PropTypes.number,
  })).isRequired,
  rowId: PropTypes.string.isRequired,
  columnId: PropTypes.string.isRequired,
};

export default Inntekt;
