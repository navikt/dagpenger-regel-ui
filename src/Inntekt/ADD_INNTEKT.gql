mutation($periode: YearMonthInput, $beloep: Float, $fordel: String, $dato: String, $beskrivelse: String) {
    addPostering(periode: $periode, beloep: $beloep, fordel: $fordel, dato: $dato, beskrivelse: $beskrivelse) {
      id
    }
  }
