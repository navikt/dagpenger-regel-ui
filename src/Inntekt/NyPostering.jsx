import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { Form, withFormik, Field } from 'formik';
import { Hovedknapp, Knapp } from 'nav-frontend-knapper';
import { Undertittel } from 'nav-frontend-typografi';
import * as Yup from 'yup';
import { InputField, SelectField } from '../Form';
import DatoLabel from '../Components/DatoLabel';
import { MMMM_YYYY_FORMAT } from '../Utils/datoFormat';
import { VerdikoderContext } from '../Context/Verdikoder';
import { ReactComponent as GodkjentIkon } from '../images/innvilget_valgt.svg';
import Spacer from '../Components/Spacer';

const mapVerdikoder = typer =>
  typer.map(({ navn }) => (
    <option value={navn} key={navn}>
      {navn}
    </option>
  ));

const visNavnOgIdentifikator = virksomhet => {
  const { navn, identifikator } = virksomhet;
  return `${navn} (${identifikator})`;
};

const NyPostering = props => {
  const { handleSubmit, isSubmitting, closeModal, values, isValid, dato } = props;
  const verdikoder = useContext(VerdikoderContext);
  return (
    <Form onSubmit={handleSubmit}>
      <div className="okavbrytmodal">
        <div className="flex">
          <div className="flexcolumn">
            <div className="okavbrytmodal--ikon">
              <GodkjentIkon />
            </div>
          </div>
          <div className="flexcolumn okavbrytmodal--text">
            <Undertittel>
              {`Legg til ny inntektspost for ${visNavnOgIdentifikator(values.virksomhet)} i `}
              <DatoLabel dato={dato} datoFormat={MMMM_YYYY_FORMAT} />
            </Undertittel>
          </div>
        </div>
        <Spacer sixteenPx />
        <div className="w400">
          <Field component={SelectField} bredde="xl" label="Beskrivelse" name="verdikode">
            {mapVerdikoder(verdikoder)}
          </Field>

          <Field component={InputField} label="BelÃ¸p" name="beloep" type="number" />
        </div>
        <Spacer sixteenPx />

        <div className="flex knapprad flexend">
          <Hovedknapp htmlType="submit" onClick={handleSubmit} disabled={!isValid || isSubmitting}>
            Legg til
          </Hovedknapp>
          <Knapp htmlType="button" onClick={closeModal}>
            Avbryt
          </Knapp>
        </div>
      </div>
    </Form>
  );
};

NyPostering.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  values: PropTypes.shape().isRequired,
  dato: PropTypes.string.isRequired,
  closeModal: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  isValid: PropTypes.bool.isRequired,
};

export default withFormik({
  mapPropsToValues: ({ arbeidsgiver, dato }) => {
    return {
      fordel: '',
      beloep: '0.00',
      dato,
      virksomhet: {
        navn: arbeidsgiver.navn,
        organisasjonsnummer: arbeidsgiver.organisasjonsnummer,
      },
      opplysningspliktig: {
        navn: arbeidsgiver.navn,
        organisasjonsnummer: arbeidsgiver.organisasjonsnummer,
      },
    };
  },

  validationSchema: Yup.object().shape({
    beloep: Yup.number().required('Required'),
    verdikode: Yup.string().required('Required'),
  }),

  handleSubmit: (values, { setSubmitting, props }) => {
    const { closeModal, arrayHelpers } = props;
    arrayHelpers.push({
      fordel: values.fordel,
      beloep: values.beloep,
      utbetaltIMaaned: values.dato,
      verdikode: 'denne kommer',
      inntektskilde: 'dagpenger-regel-ui',
      inntektsperiodetype: 'Maaned',
      virksomhet: values.virksomhet,
      inntektsmottaker: arrayHelpers.form.values.person.identifikator,
      opplysningspliktig: values.virksomhet,
    });
    setSubmitting(false);
    closeModal();
  },

  displayName: 'NyPostering',
})(NyPostering);
