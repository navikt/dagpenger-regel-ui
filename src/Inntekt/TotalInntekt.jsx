import React from 'react';
import PropTypes from 'prop-types';
import { Ingress } from 'nav-frontend-typografi';
import { formatertPengesum } from '../Utils/currencyUtils';

const TotalInntekt = ({ total }) => <Ingress>{formatertPengesum(total)}</Ingress>;

TotalInntekt.propTypes = {
  total: PropTypes.number.isRequired,
};

export default TotalInntekt;
