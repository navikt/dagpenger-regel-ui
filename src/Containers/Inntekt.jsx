import React, { useState, useContext } from 'react';
import PropTypes from 'prop-types';
import Ekspanderbartpanel from 'nav-frontend-ekspanderbartpanel';
import Modal from 'nav-frontend-modal';
import { Flatknapp } from 'nav-frontend-knapper';
import { FieldArray } from 'formik';
import SelectField from '../Components/SelectField';
import InputField from '../Components/InputField';
import { required } from '../Utils/validering';
import { OkAvbrytModal } from '../Components/OkAvbrytModal';
import { formatertPengesum } from '../Utils/currencyFormat';
import { VerdikoderContext } from '../Context/Verdikoder';
import { ReactComponent as SlettIkon } from '../images/slett.svg';


import NyInntekt from './NyInntekt';


const mapVerdikoder = typer => typer
  .map(navn => (<option value={navn} key={navn}>{navn}</option>));

// eslint-disable-next-line max-len
const sumInntekter = inntekter => inntekter.reduce((acc, val) => Number(acc) + Number(val.beloep), 0);

const inntektStyle = (rowId, columnId) => `
.inntekter--${rowId}--${columnId} {
  grid-area: inntekter--${rowId}--${columnId};
}
`;

// TODO lage egen komponent for innholdet i modal og egen for radio
const Inntekt = ({
  readOnly, inntekter, virksomhet, columnId, monthIndex,
}) => {
  const rowId = virksomhet.identifikator;
  const [editMode, setEditMode] = useState(true);
  const [isNyInntektModalOpen, setNyInntektModal] = useState(false);
  const [isSlettInntektModalOpen, setSlettInntektModal] = useState(false);
  const verdikoder = useContext(VerdikoderContext);
  return (
    <>
      <style dangerouslySetInnerHTML={{ // eslint-disable-line react/no-danger
        __html: inntektStyle(rowId, columnId),
      }}
      />
      <div className={`item inntekter inntekter--${rowId}--${columnId}`}>
        <Ekspanderbartpanel
          tittel={formatertPengesum(sumInntekter(inntekter
            .filter(inntekt => inntekt.virksomhet.identifikator === rowId)))}
          tittelProps="ingress"
        >
          <FieldArray
            name={`inntekt.arbeidsInntektMaaned[${monthIndex}].arbeidsInntektInformasjon.inntektListe`}
            render={arrayHelpers => (
              <>
                {inntekter.length > 0 && inntekter
                  .map((inntekt, index) => (
                    inntekt.virksomhet.identifikator === virksomhet.identifikator && (
                      // eslint-disable-next-line react/no-array-index-key
                      <div key={`${inntekt.verdikode}${index}`} className="flex inntekt">
                        <div>
                          <SelectField
                            bredde="xl"
                            label=""
                            selectValues={mapVerdikoder(verdikoder)}
                            validate={required}
                            name={`inntekt.arbeidsInntektMaaned[${monthIndex}].arbeidsInntektInformasjon.inntektListe[${index}].verdikode`}
                            readOnly={readOnly || editMode}
                          />
                          <InputField
                            label=""
                            name={`inntekt.arbeidsInntektMaaned[${monthIndex}].arbeidsInntektInformasjon.inntektListe[${index}].beloep`}
                            type="number"
                            validate={required}
                            formater
                            readOnly={readOnly || editMode}
                          />
                        </div>
                        {!readOnly && !editMode && (
                        <div className="flexend">
                          <button
                            type="button"
                            className="ikon ikon--slett"
                            onClick={() => setSlettInntektModal(index)}
                            title="Slett inntekt"
                          >
                            <SlettIkon />
                          </button>

                          <OkAvbrytModal
                            isOpen={isSlettInntektModalOpen === index}
                            text={`Er du sikker pÃ¥ at du vil slette ${inntekt.verdikode} - ${formatertPengesum(inntekt.beloep)}`}
                            avbrytCallback={() => setSlettInntektModal(false)}
                            OkCallback={() => {
                              arrayHelpers.remove(index);
                              setSlettInntektModal(!isSlettInntektModalOpen);
                            }}
                          />
                        </div>
                        )}
                      </div>
                    )
                  ))}
                {!readOnly && (
                  <div className="knapprad midstill">
                    <Flatknapp
                      htmlType="button"
                      mini
                      onClick={() => setEditMode(!editMode)}
                    >
                      Rediger
                    </Flatknapp>
                    <Flatknapp
                      htmlType="button"
                      mini
                      onClick={() => setNyInntektModal(!isNyInntektModalOpen)}
                    >
                      Legg til inntekt
                    </Flatknapp>
                    <Modal
                      isOpen={isNyInntektModalOpen}
                      onRequestClose={() => setNyInntektModal(false)}
                      closeButton={false}
                      contentLabel="Ny inntektspost"
                      ariaHideApp={false}
                    >
                      <NyInntekt
                        virksomhet={virksomhet}
                        dato={columnId}
                        arrayHelpers={arrayHelpers}
                        closeModal={() => setNyInntektModal(false)}
                      />

                    </Modal>

                  </div>
                )}

              </>
            )
        }
          />
        </Ekspanderbartpanel>
      </div>
    </>
  );
};

Inntekt.propTypes = {
  inntekter: PropTypes.arrayOf(PropTypes.shape({
    header: PropTypes.string,
    label: PropTypes.string,
    beloep: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  })).isRequired,
  readOnly: PropTypes.bool.isRequired,
  virksomhet: PropTypes.shape().isRequired,
  columnId: PropTypes.string.isRequired,
  monthIndex: PropTypes.number.isRequired,
};

export default Inntekt;
