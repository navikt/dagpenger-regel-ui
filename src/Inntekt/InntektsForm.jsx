import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { withFormik, FieldArray } from 'formik';
import isEqual from 'lodash.isequal';
// import { useMutation } from '@apollo/react-hooks';
// import { loader } from 'graphql.macro';
import AlertStripe from 'nav-frontend-alertstriper';
import { Element, Undertekst } from 'nav-frontend-typografi';
import Modal from 'nav-frontend-modal';
import { Hovedknapp, Knapp, Flatknapp } from 'nav-frontend-knapper';
import DisplayFormikState from '../Utils/formikUtils';
import Spacer from '../Components/Spacer';
import Inntekt from './Inntekt';
import Arbeidsgiver from '../Arbeidsgiver/Arbeidsgiver';
import NyArbeidsgiver from '../Arbeidsgiver/NyArbeidsgiver';
import OkAvbrytModal from '../Components/OkAvbrytModal';
import DatoLabel from '../Components/DatoLabel';
import { MMMM_YYYY_FORMAT } from '../Utils/datoFormat';
import { lagreInntekt } from '../lib/inntektApi';

const scrollToRef = ref => {
  const elem = document.getElementById('grid');
  elem.scrollTo({
    top: 0,
    left: ref.current.offsetLeft - 250, // 250 is the width of 1 cell
    behavior: 'smooth',
  });
};

/*
const ADD_INNTEKT = loader('./ADD_INNTEKT.gql');
const DELETE_INNTEKT = loader('./DELETE_INNTEKT.gql');
const UPDATE_INNTEKT = loader('./UPDATE_INNTEKT.gql');
*/

const InntektsForm = props => {
  // const [inntektMutation] = useMutation(ADD_INNTEKT);
  const [isArbeidsgiverModalOpen, setArbeidsgiverModal] = useState(false);
  const [isBekreftModalOpen, setBekreftModal] = useState(false);

  const { hentInntektStatus, måneder, values, isValid, dirty, readOnly, handleSubmit, status, errors, isSubmitting } = props;

  const { person } = values;

  const bekreftManuellEndring = () => {
    if (values.redigertAvSaksbehandler) {
      return setBekreftModal(true);
    }

    return handleSubmit();
  };

  /*
  const handleMutation = async () => {
    const result = await inntektMutation();

    return result;
  };

  */

  const tdRefs = {};
  const executeScroll = index => scrollToRef(tdRefs[index]);

  // Bare kjør onLoad
  if (!dirty) {
    window.setTimeout(() => {
      const elem = document.getElementById('grid');
      if (elem) {
        elem.scrollLeft = elem.scrollWidth;
      }
    }, 1);
  }

  return (
    <>
      {errors.name && <div className="error">{errors.name}</div>}
      <div className="flex hentinntekter">
        <div className="flexend">
          <Flatknapp mini htmlType="button" onClick={() => executeScroll(0)}>
            {'<<'}
          </Flatknapp>
          <Flatknapp mini htmlType="button" onClick={() => executeScroll(35)}>
            {'>>'}
          </Flatknapp>
        </div>
      </div>

      {status && status.success && (
        <div aria-live="polite">
          <AlertStripe type="suksess">
            <Element>Inntekt er lagret.</Element>
            Husk å beregn reglene på nytt i Arena slik at de inntektene du lagret nå blir med i beregningene.
          </AlertStripe>
          <Spacer sixteenPx />
        </div>
      )}

      <form onSubmit={handleSubmit}>
        <div>
          <div className="grid" id="grid">
            <div className="inntektstabell">
              <div className="row noprint">
                <div className="item" />
                {måneder &&
                  måneder.map((måned, index) => {
                    tdRefs[index] = React.createRef();
                    return (
                      <React.Fragment key={måned}>
                        <div className="item maaned" ref={tdRefs[index]}>
                          <DatoLabel dato={måned} datoFormat={MMMM_YYYY_FORMAT} />
                        </div>
                        {(index + 1) % 12 === 0 && (
                          <div className="item maaned" ref={tdRefs[index]}>
                            Totalt
                          </div>
                        )}
                      </React.Fragment>
                    );
                  })}
              </div>
              {person.vedtak.inntekt.posteringer &&
                person.vedtak.inntekt.posteringer.map((arbeidsgiver, arbeidsgiverIndex) => {
                  const key = arbeidsgiver.organisasjonsnummer || arbeidsgiver.naturligIdent;
                  return (
                    <div className="row" key={key}>
                      <Arbeidsgiver readOnly={readOnly} arbeidsgiver={arbeidsgiver} />
                      <Inntekt readOnly={readOnly} arbeidsgiver={arbeidsgiver} arbeidsgiverIndex={arbeidsgiverIndex} />
                    </div>
                  );
                })}
            </div>
          </div>
          <div className="toolbar flex knapprad">
            <div className="leggtilarbeidsgiver">
              <Knapp htmlType="button" disabled={readOnly} onClick={() => setArbeidsgiverModal(!isArbeidsgiverModalOpen)}>
                Legg til arbeidsgiver
              </Knapp>
              <Modal
                isOpen={isArbeidsgiverModalOpen}
                onRequestClose={() => setArbeidsgiverModal(false)}
                closeButton={false}
                contentLabel="Legg til arbeidsgiver"
                ariaHideApp={false}
              >
                <FieldArray
                  name="person.vedtak.inntekt.posteringer"
                  initialValues={{}}
                  render={arrayHelpers => (
                    <NyArbeidsgiver person={person} måneder={måneder} arrayHelpers={arrayHelpers} closeModal={() => setArbeidsgiverModal(false)} />
                  )}
                />
              </Modal>
            </div>
            <div className="flexend flex">
              <div className="w200 marginhoyre16">
                <Undertekst>Du må bekrefte at de nye opplysningene skal benyttes.</Undertekst>
              </div>
              <Hovedknapp
                htmlType="button"
                onClick={bekreftManuellEndring}
                spinner={isSubmitting}
                autoDisableVedSpinner
                disabled={(!hentInntektStatus && !dirty) || !isValid}
              >
                Bekreft
              </Hovedknapp>
              <OkAvbrytModal
                isOpen={isBekreftModalOpen}
                text="Når du bekrefter ny inntekt så vil alle tidligere endringene overskrives."
                avbrytCallback={() => setBekreftModal(false)}
                OkCallback={() => {
                  handleSubmit();
                  setBekreftModal(false);
                }}
              />
            </div>
          </div>
        </div>
        <div className="noprint">
          <DisplayFormikState {...values} />
        </div>
      </form>
    </>
  );
};

InntektsForm.propTypes = {
  readOnly: PropTypes.bool.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  errors: PropTypes.shape().isRequired,
  isValid: PropTypes.bool.isRequired,
  dirty: PropTypes.bool.isRequired,
  values: PropTypes.shape().isRequired,
  hentInntektStatus: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,
  status: PropTypes.shape(),
  måneder: PropTypes.arrayOf(PropTypes.string).isRequired,
};
InntektsForm.defaultProps = {
  status: undefined,
};

export default withFormik({
  mapPropsToValues: props => {
    const { initialValues } = props;
    return initialValues;
  },
  handleSubmit: (values, formProps) => {
    // console.log(values, formProps);
    const dirty = !isEqual(formProps.props.initialValues, values);

    // TODO ønsker å bruke mutation, men backend har ikke støtte for dette ennå. Så må sende inn som inntekt
    const posteringer = values.person.vedtak.inntekt.posteringer.flatMap(virksomhet =>
      Object.keys(virksomhet.posteringer).flatMap(dato => virksomhet.posteringer[dato]),
    );

    // todo fiks navn
    const groupBy = list => {
      const c = new Map();
      list.forEach(a => {
        const samme = list.filter(b => a.aarMaaned === b.aarMaaned);
        c.set(a.aarMaaned, {
          aarMaaned: a.aarMaaned,
          arbeidsInntektInformasjon: {
            inntektListe: [...samme],
          },
        });
      });
      return Array.from(c.values());
    };

    const g = groupBy(posteringer);

    if (dirty) {
      // mutate posteringer, slettet, nye og endrede
      return true;
    }
    try {
      // transform values, brukt GET values
      const inntekt = {
        inntektId: {
          id: formProps.props.initialValue.person.vedtak.inntekt.id,
        },
        inntekt: {
          arbeidsInntektMaaned: [...g],
        },
        inntektsmottaker: {
          pnr: formProps.props.initialValue.person.naturligIdent,
          navn: formProps.props.initialValue.person.navn,
        },
        redigertAvSaksbehandler: formProps.props.initialValue.person.vedtak.inntekt.redigertAvSaksbehandler,
        manueltRedigert: formProps.props.initialValue.person.vedtak.inntekt.manueltRedigert,
        timestamp: formProps.props.initialValue.person.vedtak.inntekt.timestamp,
      };

      lagreInntekt({ ...inntekt, redigertAvSaksbehandler: dirty || values.manueltRedigert }, formProps.props.hentInntektStatus, formProps.props.locationData);

      formProps.setStatus({ success: true });
      formProps.setSubmitting(false);
    } catch (error) {
      formProps.setStatus({ failure: true });
      formProps.setError(error);
      throw new Error(error);
    }

    return false;

    // console.log(dirty);
  },
  displayName: 'InntektsForm',
  enableReinitialize: true,
})(InntektsForm);
