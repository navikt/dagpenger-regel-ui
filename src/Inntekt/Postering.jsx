import React, { useState, useContext } from 'react';
import PropTypes from 'prop-types';
import Ekspanderbartpanel from 'nav-frontend-ekspanderbartpanel';
import { FieldArray } from 'formik';
import Modal from 'nav-frontend-modal';
import { Flatknapp } from 'nav-frontend-knapper';
import { InputField, SelectField } from '../Form';
import { required } from '../Utils/validering';
import { VerdikoderContext } from '../Context/Verdikoder';
import { OkAvbrytModal } from '../Components/OkAvbrytModal';
import NyPostering from './NyPostering';

import { formatertPengesum } from '../Utils/currencyUtils';
import { ReactComponent as SlettIkon } from '../images/slett.svg';

const mapVerdikoder = typer =>
  typer.map(({ navn }) => (
    <option value={navn} key={navn}>
      {navn}
    </option>
  ));

// eslint-disable-next-line max-len
const sumInntekter = inntekter => inntekter.reduce((acc, val) => Number(acc) + Number(val.beloep), 0);

const Postering = ({ arbeidsgiver, arbeidsgiverIndex, dato, readOnly }) => {
  const [editMode, setEditMode] = useState(true);
  const [isNyInntektModalOpen, setNyInntektModal] = useState(false);
  const [isSlettInntektModalOpen, setSlettInntektModal] = useState(false);
  const verdikoder = useContext(VerdikoderContext);

  const { posteringer } = arbeidsgiver;

  return (
    <Ekspanderbartpanel tittel={formatertPengesum(sumInntekter(arbeidsgiver.posteringer[dato]))} tittelProps="ingress">
      <FieldArray
        name={`person.vedtak.inntekt.posteringer[${arbeidsgiverIndex}].posteringer[${dato}]`}
        render={arrayHelpers => (
          <>
            {posteringer[dato].length > 0 &&
              posteringer[dato].map((inntekt, index) => (
                <div key={`${inntekt.fordel}${arbeidsgiverIndex}`} className="flex inntekt">
                  <div>
                    <SelectField
                      bredde="xl"
                      label=""
                      selectValues={mapVerdikoder(verdikoder)}
                      validate={required}
                      name={`person.vedtak.inntekt.posteringer[${arbeidsgiverIndex}].posteringer[${dato}][${index}].fordel`}
                      readOnly={readOnly || editMode}
                    />
                    <InputField
                      label=""
                      name={`person.vedtak.inntekt.posteringer[${arbeidsgiverIndex}].posteringer[${dato}][${index}].beloep`}
                      type="number"
                      validate={required}
                      readOnly={readOnly || editMode}
                    />
                  </div>
                  {!readOnly && !editMode && (
                    <div className="flexend">
                      <button type="button" className="ikon ikon--slett" onClick={() => setSlettInntektModal(index)} title="Slett inntekt">
                        <SlettIkon />
                      </button>

                      <OkAvbrytModal
                        isOpen={isSlettInntektModalOpen === index}
                        text={`Er du sikker pÃ¥ at du vil slette ${inntekt.fordel} - ${formatertPengesum(inntekt.beloep)}`}
                        avbrytCallback={() => setSlettInntektModal(false)}
                        OkCallback={() => {
                          arrayHelpers.remove(index);
                          setSlettInntektModal(!isSlettInntektModalOpen);
                        }}
                      />
                    </div>
                  )}
                </div>
              ))}
            {!readOnly && (
              <div className="knapprad midstill">
                <Flatknapp htmlType="button" mini onClick={() => setEditMode(!editMode)}>
                  Rediger
                </Flatknapp>
                <Flatknapp htmlType="button" mini onClick={() => setNyInntektModal(!isNyInntektModalOpen)}>
                  Legg til inntekt
                </Flatknapp>
                <Modal
                  isOpen={isNyInntektModalOpen}
                  onRequestClose={() => setNyInntektModal(false)}
                  closeButton={false}
                  contentLabel="Ny inntektspost"
                  ariaHideApp={false}
                >
                  <NyPostering arbeidsgiver={arbeidsgiver} dato={dato} arrayHelpers={arrayHelpers} closeModal={() => setNyInntektModal(false)} />
                </Modal>
              </div>
            )}
          </>
        )}
      />
    </Ekspanderbartpanel>
  );
};

Postering.propTypes = {
  arbeidsgiver: PropTypes.shape().isRequired,
  arbeidsgiverIndex: PropTypes.number.isRequired,
  dato: PropTypes.string.isRequired,
  readOnly: PropTypes.bool.isRequired,
};

export default Postering;
