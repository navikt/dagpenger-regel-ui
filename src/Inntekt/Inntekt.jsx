import React from 'react';
import PropTypes from 'prop-types';
import Postering from './Postering';
import TotalInntekt from './TotalInntekt';
import DatoLabel from '../Components/DatoLabel';
import { MMMM_YYYY_FORMAT } from '../Utils/datoFormat';

const sortObject = object => {
  const array = [];
  let prop;
  // eslint-disable-next-line no-restricted-syntax
  for (prop in object) {
    // eslint-disable-next-line no-prototype-builtins
    if (object.hasOwnProperty(prop)) {
      array.push({
        key: prop,
        totalForDato: object[prop].reduce((acc, val) => Number(acc) + Number(val.beloep), 0),
      });
    } else {
      array.push({});
    }
  }
  array.sort((a, b) => a.key.localeCompare(b.key));
  return array; // returns array
};

const sumInntekter = (inntekter, index) => {
  const sortedInntekter = sortObject(inntekter);
  // først sortere inntekter så hente ut verdier
  return sortedInntekter.slice(index - 11, index + 1).reduce((acc, val) => Number(acc) + Number(val.totalForDato), 0);
};

const Inntekt = ({ arbeidsgiver, arbeidsgiverIndex, readOnly }) =>
  Object.keys(arbeidsgiver.posteringer)
    .sort()
    .map((dato, index) => {
      return (
        <React.Fragment key={`${arbeidsgiver}${dato}`}>
          <div className="item maaned bareprint">
            <DatoLabel dato={dato} datoFormat={MMMM_YYYY_FORMAT} />
          </div>
          <div className="item inntekter">
            <Postering readOnly={readOnly} arbeidsgiver={arbeidsgiver} dato={dato} arbeidsgiverIndex={arbeidsgiverIndex} />
          </div>
          {(index + 1) % 12 === 0 && (
            <div className="item inntekter total">
              <TotalInntekt total={sumInntekter(arbeidsgiver.posteringer, index)} />
            </div>
          )}
        </React.Fragment>
      );
    });

Inntekt.propTypes = {
  arbeidsgiver: PropTypes.shape().isRequired,
  arbeidsgiverIndex: PropTypes.number.isRequired,
  readOnly: PropTypes.bool.isRequired,
};

export default Inntekt;
