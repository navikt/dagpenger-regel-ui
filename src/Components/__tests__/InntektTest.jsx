import React from 'react';
import Inntekt from '../Inntekt';
import renderer from 'react-test-renderer';

const fs = require('fs');
const path = require('path');


test('No inntekt', () => {
  const component = renderer.create(
    <Inntekt inntekter={[]} readOnly={false} columnId={'blabla'} rowId={'foofoo'} monthIndex={1} formProps={{}}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


test('Sum inntekt correctly', () => {

  let inntekt = JSON.parse(fs.readFileSync(path.resolve(__dirname, 'inntektliste.json'), 'utf8'));
  const component = renderer.create(
    <Inntekt inntekter={inntekt.inntektListe} readOnly={false} columnId={'1111111'} rowId={'1111111'} monthIndex={1}
             formProps={{}}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
