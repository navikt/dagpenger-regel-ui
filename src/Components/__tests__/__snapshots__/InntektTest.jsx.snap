// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`No inntekt 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "
.inntekter--foofoo--blabla {
  grid-area: inntekter--foofoo--blabla;
}
",
      }
    }
  />,
  <div
    className="item inntekter inntekter--foofoo--blabla"
  >
    <div
      className="ekspanderbartPanel ekspanderbartPanel--lukket"
    >
      <button
        aria-expanded={false}
        className="ekspanderbartPanel__hode"
        onClick={[Function]}
        onKeyDown={[Function]}
        type="button"
      >
        <div
          className="ekspanderbartPanel__flex-wrapper"
        >
          <span
            className="typo-element ekspanderbartPanel__heading"
          >
            kr 0,00
          </span>
          <span
            className="ekspanderbartPanel__indikator"
          />
        </div>
      </button>
    </div>
  </div>,
]
`;

exports[`Sum inntekt correctly 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "
.inntekter--1111111--1111111 {
  grid-area: inntekter--1111111--1111111;
}
",
      }
    }
  />,
  <div
    className="item inntekter inntekter--1111111--1111111"
  >
    <div
      className="ekspanderbartPanel ekspanderbartPanel--lukket"
    >
      <button
        aria-expanded={false}
        className="ekspanderbartPanel__hode"
        onClick={[Function]}
        onKeyDown={[Function]}
        type="button"
      >
        <div
          className="ekspanderbartPanel__flex-wrapper"
        >
          <span
            className="typo-element ekspanderbartPanel__heading"
          >
            kr 89 743,60
          </span>
          <span
            className="ekspanderbartPanel__indikator"
          />
        </div>
      </button>
    </div>
  </div>,
]
`;
